### ISSUE TITLE
Config (r)evolution

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#499]
Having the config in an `.ini`-like format might be a bad idea. Things like per-domain configs would probably be a lot easier in YAML...

Some links:
- https://en.wikipedia.org/wiki/YAML
- http://pyyaml.org/wiki/PyYAML
- http://yaml.org/

---

See [this proposal](https://github.com/qutebrowser/qutebrowser/issues/499#issuecomment-204430077) for more information about what the exact plans are for this.

---

~~**Note:** There's currently a [Kickstarter campaign](
https://www.kickstarter.com/projects/the-compiler/qutebrowser-v10-with-per-domain-settings?ref=3kvxwp) running with the goal of finally making this happen!~~

**Comment by pedromj** [comment]
I think this is a good idea but you have to "budget" the effort and weight it against the benefits.

**Comment by avelino** [comment]
+1

**Comment by The-Compiler** [comment]
I believe the effort is neglible if you put it in perspective with how many hours I've put into qutebrowser so far, and how many more I probably will.  I still believe in using the best tool available for the task at hand - and I think now I realized there's a better tool, and it's much more future-proof for the config.

I started the `sessions` branch (see #12) and I decided I'll use yaml for the session files, as using ini for that is definitely too cumbersome. I'll get more familiar with yaml/PyYAML there, and if it feels right, I'll convert the config stuff to yaml as well after sessions are done.

**Comment by traverseda** [comment]
I personally like prettified json. It's generally not too painful to edit, it's dead simple, and if you use it via an import then it's a pretty good place to implement a bit of custom deployment logic. Instead of having settings.yaml or settings.json, you have a settings.py with some inline json.

It works fine for projects like django, scrapy, awesomewm... Really it depends on your target audience. If you're expecting mainly programmers, then I wouldn't expect there to be any problem with a .py config file, and you'd probably get clearer error messages.

I wouldn't mind taking a crack it it if you think it's a viable direction going forward.

**Comment by The-Compiler** [comment]
> I personally like prettified json. It's generally not too painful to edit, it's dead simple, and if you use it via an import then it's a pretty good place to implement a bit of custom deployment logic. Instead of having settings.yaml or settings.json, you have a settings.py with some inline json.
> 
> It works fine for projects like django, scrapy, awesomewm... Really it depends on your target audience. If you're expecting mainly programmers, then I wouldn't expect there to be any problem with a .py config file, and you'd probably get clearer error messages.
> 
> I wouldn't mind taking a crack it it if you think it's a viable direction going forward.

I still want the config to be editable inside the browser using `:set`, so that's not really an option, as this would definitely be a config qutebrowser can't write by itself.

I already settled on using yaml for sessions (#12) so I think using yaml for the config is what I'll do.

**Comment by traverseda** [comment]
Makes sense to me.

**Comment by The-Compiler** [comment]
It might also make sense to add some dependency to validate the structure of the yaml files.

Possibilities:

* [schema](https://pypi.python.org/pypi/schema/) - cool but not commonly packaged it seems...
* [jsonschema](https://pypi.python.org/pypi/jsonschema/) - packaged for [Debian](https://packages.debian.org/search?keywords=python3-jsonschema&searchon=names&suite=all&section=all), [Ubuntu](http://packages.ubuntu.com/search?keywords=python3-jsonschema&searchon=names&suite=all&section=all), [Gentoo](http://packages.gentoo.org/package/dev-python/jsonschema?arches=all&full_cat) and probably more. Should also work for yaml.

**Comment by The-Compiler** [comment]
Retargeting this to v0.3 as I really want to get v0.2 out at some point...

Some ideas about the general structure:

```yaml
config:
    foo: bar
    baz:
    - value
    - with
    - list
    - type
    ...

keys:
    d: close
    ...

domain http://www.qutebrowser.org/:
    config:
        foo: fish
    keys:
        bar: fish
```

Then we could also easily split the config into multiple files. It probably makes most sense to have a `config.yml`, `keys.yml` and `perdomain.yml` or so.

**Comment by Ram-Z** [comment]
What about using a list for per-domain settings?

```yaml
domains:
- qutebrowser.org:
    config:
      foo: fish
    keys:
      bar: fish
- .*\.google\.com:
    config: { foo: baz }
```

**Comment by Ram-Z** [comment]
I would prefer separate files.

**Comment by angelic-sedition** [comment]
YAML makes sense for sessions, but for configuration, I'd much prefer to be able to have a single init file that just contains qutebrowser commands. This is the way most interactive applications I am familiar with do things. Mutt, pentadactyl, vim (basically), and the shell are like this for example.

For one, it's consistent. `bind` would be used instead of the keys.conf format. `quickmark-add` and `set` would be used in the config file exactly as they would be used from the qutebrowser cmd mode. I'm not sure I see the benefit of an ini or yaml config file in the first place. Maybe it's easier (or prettier to look at for some) when creating the config file if the user modifies something like `qute:settings`, but even if YAML becomes the default, I think it would be nice if there was some way to have a file of qutebrowser commands sourced on initialization.

As for per-domain settings, pentadactyl does this with "groups." Every mapping is normally in the default group. After a group is defined with the `group` command, subsequent mappings are a part of that group (a specific domain) unless otherwise specified. `-group` can also be used explicitly with commands. The reason I bring this up is that I don't like the idea of having incompatability between what can be done in the initialization config and from a running qutebrowser. Is the plan to make per-domain settings/mappings tied exclusively to a yaml config so that they cannot be created from the qutebrowser commandline?

**Comment by The-Compiler** [comment]
Thanks for your input!

> Maybe it's easier (or prettier to look at for some) when creating the config file if the user modifies something like `qute:settings`

This was the main motiviation indeed. With a config like you describe it, I can only think of these scenarios:

* The config gets overwritten when the user uses the GUI to set something. Clearly not a good thing.
* There's a separate file for the things set interactively, and it gets sourced before the "real" config. This might work...

I have to think a bit more about that idea though - coming from dwb (which does something similiar to what qutebrowser does now) I didn't even consider this. I definitely see the advantages.

I think there are some other caveats though, especially related to interactive settings and escaping of chars.

> I think it would be nice if there was some way to have a file of qutebrowser commands sourced on initialization.

That's planned as a part of #35. That means even if I go for YAML, you could still not set things interactively (which keeps the file empty) and write such a file. :smile:

> Is the plan to make per-domain settings/mappings tied exclusively to a yaml config so that they cannot be created from the qutebrowser commandline?

No, I planned to add something like `--domain` to `:set`.

**Comment by The-Compiler** [comment]
I just discussed with some people in the `#archlinux` IRC channel.

It probably would make sense to have separate config files for things set by the user vs. things set in the GUI.

Then the GUI config gets auto-migration etc., and the user config never gets touched. Novice users will just use the GUI, and more experienced users will write their own config file (which is initially empty).

As for precedence, I think the user config should win, because *removing* something from the gui config after it's been added might be tricky and surprising.

**Comment by NoSuck** [comment]
That sounds like a good compromise, as the users who are more likely to edit config files in text tend to be the same users that would gripe when hand-edited configs are overridden by auto-migration.

**Comment by angelic-sedition** [comment]
> No, I planned to add something like --domain to :set.

Sounds great!

> It probably would make sense to have separate config files

I agree; that sounds perfect. I particularly like the idea of it being empty at first. In config files meant for the user, I prefer being able to only override or add to the default configuration instead of having to have all configuration/mappings (even if they haven't been altered) in the file. The suggested precedence makes sense as well.

**Comment by Carpetsmoker** [comment]
I had a wildly different idea about configuration: make all configuration in Python.

Anyone who has used Vim or Emacs understands the power that comes when you use a scripting language to configure a program. There are all sorts of 'clever things' you can do, from the fairly mundane:

	" Use blowfish2 for encrypting files; blowfish is *not* secure
	if has("cryptv") && has("patch-7.4-399")
		set cryptmethod=blowfish2
	endif

To the slightly more advanced:

	" Always make n to the next match, even when using ?
	nnoremap <expr> n 'Nn'[v:searchforward]
	nnoremap <expr> N 'nN'[v:searchforward]

And even adding entirely new features (like [showing a fortune on startup](http://code.arp242.net/config/src/8fbc0c013bc6d35a0b755f979d04a8626a656901/modules/vim/fortune.vim?at=default), or [an address book](http://code.arp242.net/config/src/8fbc0c013bc6d35a0b755f979d04a8626a656901/modules/vim/address.vim?at=default)) is not terribly complicated.

In qutebrowser, Python would be the obvious choice to do this in. As a very basic example: [one of my programs does this](
http://code.arp242.net/battray/src/d04ba45806fdd7d0875fed9bb6dc9b17005b2f59/data/battrayrc.py?at=default), the code is run with [`exec`](http://code.arp242.net/battray/src/d04ba45806fdd7d0875fed9bb6dc9b17005b2f59/battray.py?at=default#cl-109)...

**Comment by lamarpavel** [comment]
That would complicate modifying the settings from within qutebrowser and then writing the config to file.

**Comment by traverseda** [comment]
Yeah, we went down that road.

One option could be a defaults.json. Then we have a settings.py that loads the defaults, and allows for overrides.

```
settings = json.loads(defaults.json) #System defaults. These get changed with new releases.
settings.update(json.loads(userSettings.jsin)) #Settings set via the set command. Store's user settings that aren't actually in a settings file

settings['beAwesome'] = isAwesome() #Define settings via a central server, custom functions, whatever. Because your settings are actually python.
```

Of course I say json here, but it could just as easily be yaml.

**Comment by pedromj** [comment]
On Sun, May 31, 2015 at 11:19:13AM -0700, Alex Davies wrote:
> Yeah, we went down that road.
> 
> One option could be a defaults.json. Then we have a settings.py that loads the defaults, and allows for overrides.
> 
> ```
> settings = json.loads(defaults.json) #System defaults. These get changed with new releases.
> settings.update(json.loads(userSettings.jsin)) #Settings set via the set command. Store's user settings that aren't actually in a settings file
> 
> settings['beAwesome'] = isAwesome() #Define settings via a central server, custom functions, whatever. Because your settings are actually python.
> ```
> 
> Of course I say json here, but it could just as easily be yaml.

I was just going to propose an option like this. However, I prefer using
objects instead of dicts:

1. Load settings.[json|yaml] from user config dir and create the config
structure (objects).
2. Execute some "custom_settings.py" also in user config dig, providing
the config structure for the script to tune it.

Something similar to this could be fine to somewhat support "conditional
configuration", what I think is the objective.

Regards,
Pedro

-- 
Pedro Martinez-Julia
Department of Communication and Information Engineering
Faculty of Computer Science
University of Murcia
Email: pedromj@um.es
---------------------------------------------------------
*** Entia non sunt multiplicanda praeter necessitatem ***

**Comment by The-Compiler** [comment]
I really, *really* like @traverseda's idea.

Then the GUI config could be written to the .yml file(s), and by default, the "user config" is a bit of python code which loads those files.

Maximal flexibility for the user, and it's nicely defined what takes precedence. If the user doesn't care about GUI settings (or only wants them to be temporary), they just remove the code loading it.

That's genius! :smile:

**Comment by ikhthiandor** [comment]
You can also consider TOML the alternative to YAML.

https://github.com/toml-lang/toml

**Comment by ikhthiandor** [comment]
You can also consider TOML as the alternative to YAML.

https://github.com/toml-lang/toml

**Comment by ikhthiandor** [comment]
You can also consider TOML as the alternative to YAML.

https://github.com/toml-lang/tomlG

**Comment by The-Compiler** [comment]
Some discussions from IRC until I get to write a proper proposal/summary of this:

* `<The-Compiler>` I agree config categories must die, it'll probably be replaced by using dotted paths where it makes sense (`colors.statusbar.fg`, `fonts.statusbar`, etc.) and settings without dots/categories/namespaces where it makes more sense - see #893
* `<The-Compiler>` I'd really like to keep the ability to configure things from the GUI (via `:set` and/or `qute:settings`), and that'll probably land in a `guiconfig.yml` or so - while you can edit it by hand, it's not intended to be, and qutebrowser will mess with it like it sees fit
* `<The-Compiler>` In addition to that, there'll either be a `config.py`, or a `config.rc`. In that file, you'll have to explicitly load the default keybindings and the `guiconfig.yml` somehow (and the default will be exactly that, and no more), and then you can add more settings on top - either via a Python API (which also would allow easily loading more yaml files), or via that rc file. I'm open to both approaches really, not sure what'd be better yet.
* `<The-Compiler>` so e.g. your `config.rc` (if we go for that) would be `:bind --load-defaults; :set --load-guiconfig; :set blah blub; :bind xy cmd`
* `<The-Compiler>` that means someone could easily opt-out from getting the default keybindings if they want their own, completely customized keybindings
* `<craftyguy>` having the "user config" as a py would be AMAZING. then I could have a unified config file, with customizations for each of the systems I use! :P
* `<The-Compiler>` and if someone wants `:set` to never persist things at all, just turn off the loading of `guiconfig.yml`
* `<The-Compiler>` for people who don't want to edit configs by hand at all, simply don't mess with the `config.{rc,py}` and use `:set`/`qute:settings`
* `<craftyguy>` The-Compiler: preventing `guiconfig.yml` would disable ALL spooky config changes behind the scenes from qute?
* `<The-Compiler>` craftyguy: that's the point, yeah. qutebrowser only ever messes with `guiconfig.yml`
* `<Ferus>` I'm completely ok with everything that was just listed tbh, sounds great :)
* `<The-Compiler>` which is auto-migrated on updates, and auto-saved when doing `:set`
* `<The-Compiler>` and when using the `config.{rc,py}`, none of that is done, in return you're responsible of upgrading your config
* `<craftyguy>` The-Compiler: "you're responsible.." will we get warnings if a deprecated option is detected? I'm guessing that if you add a new option to the browser and if it doesn't detect it in the user config, it'll just use some default so it won't crash.. correct?
* `<The-Compiler>` yeah, like if you'd run :set with wrong parameters
* `<The-Compiler>` the reason it refuses to start entirely at the moment is because it'd delete that option from the config on the next save, which isn't exactly nice either

**Comment by The-Compiler** [comment]
So I've thought about this some more, and even though QtWebEngine is coming first, this is finally coming a bit closer :wink:
Dumping my thoughts here, all feedback about this would be welcome.

# Philosophy

There are essentially two different workflows (or types of users) for configs:
  - Setting things via the GUI, using `qute:settings` or `:set` and not editing the config by hand (like dwb)
  - Editing the config by hand (and e.g. managing it in a git repo), without qutebrowser writing to the config (like vimperator/vim/...)
  
I'd like to satisfy both user groups, and my first attempt at that (i.e. the current one) definitely failed :laughing:

# Basics

- qutebrowser has two config files, `config.py` and `guiconfig.yml`
- By default, those files are missing/empty, and the built-in defaults get loaded
- qutebrowser will take care of `guiconfig.yml`, but never mess with `config.py`
- config sections don't exist anymore, instead there are dotted option paths

# Configuration via GUI

- Using `:set` or `qute:settings` will write everything changed from the defaults to `guiconfig.yml`
- As soon as a setting is set back to its default, it will be deleted from `guiconfig.yml` again
  - pro: The config stays small, new defaults can easily be added as the defaults are built-in
  - contra: The user can't explicitly pin default values (but this type of user probably prefers `config.py` anyways?)
- On updates, qutebrowser will auto-migrate values in `guiconfig.yml`

The YAML looks like this:

```yml
config:
    foo: bar
    baz:
    - value
    - with
    - list
    - type
    ...

keys:
    d: close
    ...

domains:
    - qutebrowser.org
      config:
          foo: fish
      keys:
          bar: fish
```

# Configuration via config.py

- As soon as a `config.py` is present, the `guiconfig.yml` is not loaded anymore, as the user prefers to set things explicitly
- The `config.py` is python, and a global `config` object (a module) is available
- Optionally, this will do the same (for people who e.g. want autocompletion by their IDE/editor):

```python
from qutebrowser.config.configapi import ConfigAPI
config = ConfigAPI()
```

- The user is reponsible for updating their config on qutebrowser updates
- qutebrowser will show errors and load the defaults on errors in the config (or load the part it could parse)

## Available methods/attributes

### `config.set(setting, value, domain=None)`

Set the given `setting` to the given `value` (optionally for the given `domain` as soon as there are per-domain settings)

`value` isn't necessarily a string (like now), depending on the type of the value to set it could be:

- `bool` (`True`/`False`)
- `str` (`"foo"`)
- `float` (`1.2`)
- `int` (`23`)
- `list` (`["val1", "val2"]`)
- `dict` (`{"key": "val"}`)

Example: `config.set('general.developer-extras', True)`

### `config.bind(key, command)`

This binds a key(chain) to a command.

`key` is a string, `command` is a list of strings.

(maybe `command` should be a string, split like in the qutebrowser commandline?)

Example: `config.bind('x', ['spawn', 'mpv', '{url}'])``

### `config.load_default_bindings`

An attribute, set to `True` by default. You could do
`config.load_default_bindings = False` to avoid loading any of the default
keybindings. This is for people who want to bind a completely different keyboard scheme.

### `config.load_yaml(filename)`

Load a different yaml config file in the config dir.

This could also be used like `config.load_yaml('guiconfig.yml')` to load `guiconfig.yml` even though a `config.py` is present.

### `with config.domain('...'):`

This can be used as a Python context manager to group settings belonging to a given domain, e.g. those two snippets are equivalent:

```python
with config.domain('qutebrowser.org'):
    config.set('key1, 'val1')
    config.set('key2, 'val2')
```

```python
config.set('key1, 'val1', domain='qutebrowser.org')
config.set('key2, 'val2', domain='qutebrowser.org')
```

**Comment by lamarpavel** [comment]
Minor nitpick: Calling it `guiconfig.yaml` can be confusing - implying that it's configuration related only to the gui.
Instead I would call it something like `managedconfig.yml`.

**Comment by neeasade** [comment]
genconfig, autoconfig? -- if it's filled only things that are non-defaults, maybe customoptions/customsettings.yml?

**Comment by lamarpavel** [comment]
Under `config.bind(key, command)` you wrote:
>(maybe command should be a string, split like in the qutebrowser commandline?)

I agree. Specifically the string should be the same as in the `guiconfig.yml` under `keys:` to make things more intuitive. When I want to add a new keybinding, it's good to be able to simply look up the command in the yaml file instead of figuring out how to split the arguments in a list.

**Comment by lamarpavel** [comment]
@neeasade customoptions/customsettings.yml is a name that would just as well fit the config.py, I think we should have a name that implies
1. that it's managed by qutebrowser
2. that it's not the file you want to edit by hand
3. that it overrides defaults

autoconfig.yml sounds okay. But there should still be a comment at the top of the file stating that you shouldn't edit it by hand.

**Comment by Link-Satonaka** [comment]
Isn't this all needlessly complicated?

Use one config.
- Contains only what is _not_ default
- Implement `:source` to rehash the config on demand, `:source ~/.config/qutebrowser/config`
- qute:settings reads (`:source` upon opening the page) and writes the config. Users should be instructed to not manually edit the config while this page is open, or changes may be lost. The browser should not write to the config at any other point, excluding the explicit `:save` (and also upon exit, if the user has enabled save on exit).

Done.

The above is a syntax/implementation neutral idea, but I highly advise you follow the vim way and use rc (where qute:settings calls `:mkquterc` a la vim's `:mkvimrc`, and we ditch `:save`, or make it an alias). You place a high value on the ability to configure everything in browser via `:set` (+ other commands), yet you appear to be afraid of rc style config- the entire premise of which depends on having the ability to do just that!

Using rc will be much simpler for users to look at the config and understand it, if they so choose. After all, there is nothing to understand but the in-browser functions that should already be familiar. What you propose forces the user to learn two sets of syntaxes- the config syntax, and the in-browser syntax. You proposed `config.bind('x', ['spawn', 'mpv', '{url}'])`. While not terribly difficult... just, why? With rc, the config would be exactly as you call it in-browser `:bind x spawn mpv {url}`


I really think you're reinventing the wheel here, and you're introducing needless complexity for yourself as the developer, and the end user.

**Comment by The-Compiler** [comment]
So you seem to be talking about two different things here - let me answer to them separately:

## Using one/two config files, `:set`/`:mkquterc`

As said before, I'd like to keep the ability to set and persist settings from the browser, without the user having to explicitly `:save` them.

> qute:settings reads (`:source` upon opening the page) and writes the config

How is that going to work? How would qutebrowser write the config (*especially* in a custom python/rc format) without losing the user's structure, comments, etc.?

Also, what about migrations on update? Some qutebrowser users (i.e. the one typically using the convenience of `:set`) really don't want to care about upgrading their config by hand on every qutebrowser update.

That's just not going to work without messing things up. Better to keep those two things separate, and let the user decide if they want to use `:set`/`qute:settings` or edit a file in their editor.

## Using Python vs. rc syntax

I'm still not sure about this one. The problem I see with `rc` files is that a lot of tools using them (vim, pentadactyl, mutt, ...) end up inventing their own mini-language (or in case of vim, not-so-mini language) so you can do things like:

- comments
- `if has("somefeature")`
- use heredoc-like multiline-strings
- substitute variables
- etc. etc.

Now *that* is unneded complexity. Why invent my own (vim-like-ish?) mini-language when I just can delegate that task to Python? I'll get all that for free, and the user gets additional power (like, say, a different config based on the hostname) for free.

**Comment by Link-Satonaka** [comment]
>How is that going to work? How would qutebrowser write the config....
There are many ways you could tackle that.

- I would suggest updating lines in-place. While parsing the config, a line `:set ui.showtabs false` is easily recognized and replaced with `:set ui.showtabs true`. Optionally inject a comment `:set ui.showtabs true // modified in browser`
- rc is processed linearly from top to bottom. Just write the new settings at the end of the file. Any previous definition of the setting will be overridden.
- Pure vim style. Completely overwrite and replace the config. This assumes the user is either writing their first config, has no interest in editing the config by hand, or has versioned backups of their config (git) and wants to issue the `:save` command for whatever reason.

I agree with your assessment of rc, but I maintain that creating two sets of configs is pretty crazy, and we should work on solving the puzzle instead of trying to clumsily bypass it.

**Comment by Link-Satonaka** [comment]
I think the power of python could be easily wielded by a custom userscript, though. No need to expand your rc to anything but the literal _run command_, and one of the commands could be `:spawn -u conditionals.py`

**Comment by The-Compiler** [comment]
So something which came up is that Python might be difficult/unknown to people not used to it. I edited the [proposal](https://github.com/The-Compiler/qutebrowser/issues/499#issuecomment-204430077) to add a `config.rc` format. This will be trivial to implement, and then `config.py` is more like a startup-hook thing (with some API) for more advanced configuration.

**Comment by Eisfreak7** [comment]
I like the idea to use python for configuration (or to glue together a configuration), because it gives that much more flexibility than a plain rc file.
For me personally, the possibility to edit the config in-gui is not needed (and I suspect many qutebrowser users agree, since this kind of browser should appeal to a specifig kind of user). But regardless, at least partial programming support in the configuration support is great.

**Comment by Link-Satonaka** [comment]
I think this updated proposal is a fair compromise, as I understand it. I'd just like to clear up a few things.

- Am I correct in thinking `qute:settings` will be independent of both rc/py config? To rationalize this, there should be an option to _not_ write changes to disk at any point, and if rc/py configs are used, disable the `:save` command (or throw a `Warning: you are writing a yml config when {whateverconfig} exists!`). This way, the browser will (optionally) behave in the way rc users expect: no changes are preserved across sessions unless manually added to their rc.
- Are we married to the name `config.rc`? Where is it going? I won't oppose whatever you choose, but might I suggest both `$HOME/.qutebrowserrc` AND `$HOME/.config/qutebrowser/qutebrowserrc` . Note that it is a dotfile only when in the root of `$HOME`.
- Are you for or against the idea of a `:source` command?
- To avoid confusion, would you be opposed to having ONE manual config `.qutebrowserrc`, where qutebrowser determines what type of config it is by the first line? For example, 
```bash
#!/bin/python
config.bind('x', ['spawn', 'mpv', '{url}'])
```
or
```bash
#rc
bind x spawn mpv {url}
```
Just feels tidier this way. In the Help/FAQ you'd write something like `Your config file can be written in either rc or python!` Note that the inclusion of a `:source` function could/should enable users to have both rc and python configs (as pointless as that may ultimately be since python does everything rc does and more :smile:  ).
```bash
#rc
bind x spawn mpv {url}
source /path/to/other/config # where config is, as expected, interpreted as rc or python by the first line
```

**Comment by The-Compiler** [comment]
> I think this updated proposal is a fair compromise, as I understand it. I'd just like to clear up a few things.

Thanks for taking the time to do so!

> - Am I correct in thinking `qute:settings` will be independent of both rc/py config? To rationalize this, there should be an option to _not_ write changes to disk at any point, and if rc/py configs are used, disable the `:save` command (or throw a `Warning: you are writing a yml config when {whateverconfig} exists!`). This way, the browser will (optionally) behave in the way rc users expect: no changes are preserved across sessions unless manually added to their rc.

In the current proposal, `qute:settings` would always (and only) write to
`autosettings.yml`.

Then in your `config.py` you decide whether to load that file or not.

If you have a `autosettings.yml` and `config.rc` (but no `config.py`), the
default `config.py` of
`config.load_yaml('autoconfig.yml'); config.load_rc('config.rc')` would apply,
i.e. both would be loaded, with `config.rc` taking precedence.

> - Are we married to the name `config.rc`?

Not at all. It was just the first thing coming to mind.

> Where is it going? I won't oppose whatever you choose, but might I suggest both `$HOME/.qutebrowserrc` AND `$HOME/.config/qutebrowser/qutebrowserrc` . Note that it is a dotfile only when in the root of `$HOME`.

I'd rather not make the "where to read the config" thing more complicated. Is
there a downside to *requiring* it's in `~/.config` (as per XDG)?

> - Are you for or against the idea of a `:source` command?

For `rc` style files, sure, sounds like a good idea.

> - To avoid confusion, would you be opposed to having ONE manual config `.qutebrowserrc`, where qutebrowser determines what type of config it is by the first line? For example, 
> ```bash
> #!/bin/python
> config.bind('x', ['spawn', 'mpv', '{url}'])
> ```
> or
> ```bash
> #rc
> bind x spawn mpv {url}
> ```
> Just feels tidier this way. In the Help/FAQ you'd write something like `Your config file can be written in either rc or python!` Note that the inclusion of a `:source` function could/should enable users to have both rc and python configs (as pointless as that may ultimately be since python does everything rc does and more :smile:  ).
> ```bash
> #rc
> bind x spawn mpv {url}
> source /path/to/other/config # where config is, as expected, interpreted as rc or python by the first line
> ```

Oh - I haven't considered that yet! In the proposal above, I have viewed
`config.py` as the "glue" between the different config files (or as some kind
of startup hook python script), with the optional ability to also modify
settings from there.

Your proposal would streamline things a bit. To summarize:

- As soon as a config file is present, `guiconfig.yml` gets ignored
- The config is either rc- or python-style as identified by the shebang (or something else)
- In both formats, there'd be a way to source other config files (either rc or Python)

What do others think? I feel like I need to start tracking the proposal in the
repo :laughing:

(I really appreciate the input of everyone here!)

**Comment by Eisfreak7** [comment]
I think that'd be a nice convenience, but I wouldn't bloat the code for it too much. If its easy to fit in its great, otherwise I don't think its too much hassle to have a single line python config that sources the rc.

**Comment by noctuid** [comment]
Whatever is simplest/easiest would be fine with me as long as there can be both rc and python configs.

**Comment by lahwaacz** [comment]
> In the current proposal, qute:settings would always (and only) write to
autosettings.yml.
>
> Then in your config.py you decide whether to load that file or not.

What about `:set`? Would it also always (and only) write to autosettings.yml? Maybe I just didn't understand it, but the current proposal seems that if autosettings.yml is not loaded from config.py, then `:set` will have no effect. Maybe you'll need to detect if autosettings.yml has been loaded and otherwise make it affect the session only temporarily. This will likely add too much confusion, so maybe the persistent and temporary `:set` should be split to separate commands...

One of the few features of qutebrowser that I never quite understood is the `--restore SESSION` command line opption. For one thing, it seems that "session" is not the same as "instance", which I find very unnatural:

    $ qutebrowser -r foo
    INFO: Opening in existing instance

If the only reason is to avoid conflicts on file writes, it looses its purpose if users choose not to use autosettings.yml (I'd imagine any other path with write conflicts, like cache, could be changed to include the session name).

Now it seems that the only way to start a separate instance is to use `--basedir` or `--temp-basedir`. But what if multiple instances with the same `--basedir` are started? How are the file write conflicts solved?

**Comment by The-Compiler** [comment]
> What about `:set`? Would it also always (and only) write to autosettings.yml? Maybe I just didn't understand it, but the current proposal seems that if autosettings.yml is not loaded from config.py, then `:set` will have no effect. Maybe you'll need to detect if autosettings.yml has been loaded and otherwise make it affect the session only temporarily.

`:set` already does two things:

- Set the setting in the current session
- Save the setting to the config (or `autoconfig.yml` with the proposal)

If you're not loading `autoconfig.yml`, the settings will be temporary, even
when they're *saved* to that file.

> This will likely add too much confusion, so maybe the persistent and temporary `:set` should be split to separate commands...

There is `:set -t` / `--temp` already.

> One of the few features of qutebrowser that I never quite understood is the `--restore SESSION` command line opption. For one thing, it seems that "session" is not the same as "instance", which I find very unnatural:
> 
>     $ qutebrowser -r foo
>     INFO: Opening in existing instance
> 
> If the only reason is to avoid conflicts on file writes, it looses its purpose if users choose not to use autosettings.yml (I'd imagine any other path with write conflicts, like cache, could be changed to include the session name).

You lost me there, how is this all related to configuration? It seems like you're talking about `--basedir` (or profiles, like other browsers call the same concept and I probably should too).

Sessions are simply a dump of all your browsing state (open tabs/windows/scroll position/zoom/history). Not related to config or where files are stored at all.

> Now it seems that the only way to start a separate instance is to use `--basedir` or `--temp-basedir`. But what if multiple instances with the same `--basedir` are started? How are the file write conflicts solved?

You can't start multiple instances with the same `--basedir`, as the basedir path is encoded in the FIFO used to communicate with a potential existing instance - so if you use the same basedir, the existing instance will take over too.

**Comment by lahwaacz** [comment]
> `:set` already does two things:
> 
> - Set the setting in the current session
> - Save the setting to the config (or `autoconfig.yml` with the proposal)
> 
> If you're not loading `autoconfig.yml`, the settings will be temporary, even
> when they're *saved* to that file.

So there's actually no problem, I'm glad to see that. By the way, am I right
that `autoconfig.yml` is/will be loaded only once at start up and then serve
only as a write-only dump?

> There is `:set -t` / `--temp` already.

I must have missed that, sorry about that...

> > One of the few features of qutebrowser that I never quite understood is the `--restore SESSION` command line opption. For one thing, it seems that "session" is not the same as "instance", which I find very unnatural:
> > 
> >     $ qutebrowser -r foo
> >     INFO: Opening in existing instance
> > 
> > If the only reason is to avoid conflicts on file writes, it looses its purpose if users choose not to use autosettings.yml (I'd imagine any other path with write conflicts, like cache, could be changed to include the session name).
> 
> You lost me there, how is this all related to configuration? It seems like you're talking about `--basedir` (or profiles, like other browsers call the same concept and I probably should too).
> 
> Sessions are simply a dump of all your browsing state (open tabs/windows/scroll position/zoom/history). Not related to config or where files are stored at all.

I'm sorry for going too off-topic, the only relation was the concern about
conflicts when writing to files. I will take the rest of my
concerns/misconceptions to a separate issue...

**Comment by The-Compiler** [comment]
> So there's actually no problem, I'm glad to see that. By the way, am I right
that `autoconfig.yml` is/will be loaded only once at start up and then serve
only as a write-only dump?

Yeah, though I might add a `:source` command which can also source that file.

**Comment by t1mxg0d** [comment]
I agree with @ikhthiandor regarding TOML over YAML. It is closer to ini style structure than YAML, which will make the translation easier but with a very clean syntax. I'm starting to see TOML slowly replacing YAML in various projects.

Example: YAML vs TOML https://gist.github.com/oconnor663/9aeb4ed56394cb013a20

Either way though, YAML will be a big improvement. The most painful part of Qutebrower for me is config. Additionally, the extremely long inlined documentation comments makes it hard to quickly jump to different values/sections. But that's a different problem. :+1:

**Comment by The-Compiler** [comment]
FWIW, I find the YAML in that example more readable than TOML! I'm not saying I'm *right* though, I'm just saying it's a matter of opinion.

What I can absolutely see though is that YAML is harder to *edit*, especially with autoindentation in vim which seems to do very weird stuff there...

What keeps me back from using TOML is that it's an additional dependency (YAML is already used for session files) and it's not packaged for all the distributions qutebrowser is (or will be soon) last time I checked.

However, as per https://github.com/The-Compiler/qutebrowser/issues/499#issuecomment-204430077 the YAML file will mainly be there for settings you adjust using the GUI (`qute:settings`, `:set`, `:bind`). If you prefer to write a settings file, you'll have the possibility to do that in either Python or rc-style.

**Comment by mlochbaum** [comment]
There are two issues at hand here. The first, how to store the configuration(s), has been adequately addressed: allowing a few types of configurations is flexible, not terribly difficult, and can be implemented one filetype at a time. Anyone motivated can start working on code for it now.

The second is the organization of various configurations, and when and how they should be loaded. I think there are a lot of shortcomings in the previous proposals here. Following is a description of what our system needs to do, and a partial proposal for how to accomplish it.

## Requirements
We need at least all of the following as possible scopes for configuration.
* Default
* Global (current: `*.conf`)
* Per-session (current: `--temp`)
* Per-window
* Per-tab
* Per-site (proposed: `config.domain`)

Conceivably other scope-like things could be added:
* The current tab, and tabs opened implicitly from it
* The type of file being browsed
* The current mode (as used with keybindings)
* The exact current page (not URL; ends on any navigation)

I won't try to defend all of these here. As an example of when one of the more obscure ones, the tab-plus-implicit-children scope, is used, suppose a user wants to set `private-browsing` for a particular usage of a site. They also use this site in a non-private manner (maybe it's just a search engine), so per-domain settings don't work. Additionally, links visited from the current tab should be subject to the same rules, since they could reveal information about the original site. Setting `private-browsing` for the current tab, and tabs opened from it, accomplishes exactly what we want to do.

Additionally, the user should be able to apply these scopes to groups of options, rather than just single values. To extend the `private-browsing` example, it might be a good idea to change the color of the tab bar or status bar for sites with private browsing enabled, so the user can easily check if the current tab is private.

The global and per-site configurations should be held in files which can be edited, reloaded, and written, and it should always be possible to load a configuration from a file into a specific scope.

## Proposal
Each scope should maintain its own configuration object, which gives the options which are added or modified by that scope. These should be composed to give the actual configuration used by the browser.

Most of the scopes I gave are naturally ordered. For example the window scope should override the session scope, since otherwise any option which has a value for the session could not be customized across windows. For domains there is no clear order: typically a per-tab option is less important that a given per-site option, but it's still necessary to be able to override per-site options without changing them permanently. To avoid our hierarchy breaking down, we should split the conditions attached to a option into scopes (set an option for a certain scope) and attributes (specify, for an option, conditions under which it should be applied).

#### Scopes
1. Default
2. Global
3. Session
4. Window
5. (Optional) Tab, including implicit children
6. Tab
7. (Optional) Current page only

#### Attributes
* Site information (domain, filetype)
* Mode

An attribute is then part of an option's value, while a scope is where the option is stored.

### Syntax
The scope should be an argument, and attributes should be something flag-like. So:
```
:set window --domain=github.com colors.statusbar.bg red
```

`:load-file` or whatever we call it should have similar options.

### Other considerations
Some options can only be set for some scopes: `last-close`, for instance, really only makes sense as a per-window setting. On the other hand, setting `tabs-are-windows` per-tab only sounds paradoxical.

It's probably a good idea to be able to dump the current configuration in effect for a given tab to a file.

Keybindings and configuration should use pretty much the same style, but the configuration needs to be known first if we want to make a `use-default-keybindings` option.

**Comment by The-Compiler** [comment]
@mlochbaum I don't think this would be a good idea. While you gain maximum flexibility, this is at the cost of complexity.

If as an user I have to check ten different places to find out why the heck some option is turned on, that's not a configuration system I want to use. Of course, as a developer, this also makes configuration *a lot* more difficult to implement and understand.

I think we need to draw a line somewhere instead of trying to shoehorn everything into a config system. What does actually have multiple good use-cases and can be generalized enough, and what can/should be solved in a maybe less flexible, but much easier or more fitting way?

For per-domain settings, I can think of many possibilities - turning on/off javascript for a site, setting a specific user-stylesheet for one ("stylish" essentially), disabling adblocking for something you'd want to support, etc.

For private browsing, it might actually make more sense - also from an UI perspective - to have `:open --private` which opens the given URL in a new "icognito window" similarily to what Chromium/Firefox do and not have a setting at all? Yes, it's less flexible. But it's so much easier to understand.

Per-tab settings - maybe. I've been thinking about this one for months, and I'm not sure. I'm also not sure what settings would even be implementable per-tab with QtWebKit/QtWebEngine's requirements. It's something I'd rather keep in the back of my mind and look at in the future when there's enough need.

For other fancy things, I think the plugin API (#30) should be a better fit, so you can simply implement the hooks you want, and then change the settings there in your plugin.

**Comment by lamarpavel** [comment]
Some ideas:

Tab-local settings: Yes, but only at runtime and not managed by either `autoconfig.yaml` or `config.rc`. This should not be intended as normal user-interfacing config but rather be available to userscripts that are then responsible for management. Think of a userscript that opens a new tab, puts qutebrowser in fullscreen-mode and disables the UI and all keybingings except for navigation - reverting all of that when the tab is closed again - in a set-and-forget manner that will never be saved. The vim analogy would be plugins that open the quickfix-list, rebinding keys just for that buffer and forgetting all about it as soon as the quickfixlist is closed again.

However, treat this as convenience and nice-to-have, but don't let it increas the complexity for the important parts of the configuration (hence the not-saving policy as save/load seems the main cause of complexity here).

An issue I have not see addressed here is this:
```
The rc is intended for simple configurations by users preferring a config
file. Any non-blank non-comment line (^[^#].*) will be executed as a
qutebrowser command, with :set/:bind not writing to autoconfig.yml for the
commands executed from config.rc
```
So the set/bind commands in the rc-file are not propagated to the yaml-file. Good. Now what if my rc-file is setting `optionA` to non-default value and while qutebrowser is running I do `:save` (or whatever the command will be to save current config to the yaml-file)? Because as you said right before, when saving the config from the UI, all non-default settings are stored in the yaml.

**Comment by bb010g** [comment]
> What keeps me back from using TOML is that it's an additional dependency (YAML is already used for session files) and it's not packaged for all the distributions qutebrowser is (or will be soon) last time I checked.

The dependency issue could be resolved if you switch by also switching over the session files to TOML. You shouldn't loose anything capability-wise.

**Comment by The-Compiler** [comment]
> So the set/bind commands in the rc-file are not propagated to the yaml-file. Good. Now what if my rc-file is setting `optionA` to non-default value and while qutebrowser is running I do `:save` (or whatever the command will be to save current config to the yaml-file)? Because as you said right before, when saving the config from the UI, all non-default settings are stored in the yaml.

Hmm, good point. I guess `:set`/`:bind` commands from the `.rc` file would need to end up in some special layer which are never persisted to disk (by qutebrowser), essentially like `:set --temp` does now.

Honestly, I'm beginning to wonder again if we really need an `.rc` file, or if everyone will be able to deal with a simple Python file when it's nicely documented. Apart from making the implementation more straight forward, it also makes things easier from a UX support of view. I can imagine having two different config formats to be kind of bothersome from a documentation and support (as in helping people) standpoint too...

> The dependency issue could be resolved if you switch by also switching over the session files to TOML. You shouldn't loose anything capability-wise.

That still doesn't make TOML packaged in distributions, and I'd rather avoid being unable to read old session files.

**Comment by mlochbaum** [comment]
@The-Compiler I think you are overestimating the complexity of the proposal for the user. Default, global, and session layers are already supported by qutebrowser, and all of the other scopes are temporary. The user should be able to remember things they set in the same session, and if not restarting the browser clears everything anyway. None of these settings would be loaded automatically from files, and I would expect the temporary scopes to be empty most of the time.

Per-tab settings are pretty important in my opinion. I often find that I want to change something for the current page I am on--enable Javascript, remove the scrollbar, change the font, temporarily allow cookies--without making the same change for other tabs or for the same domain later. Setting with `--temp` and a domain will mostly work, but could show up later if I happen to visit the same domain again; it's just not what I really wanted to do.

The attribute model probably needs more work, and I'm not sure what kind of data layout works best for it. But we have to have something to store per-domain options, and it seems that there are other attribute-like things that would benefit from the same framework.

Not everything needs to be planned implemented all at once. But I think the scope-attribute framework is needed just to make sense of the options that we know we need to support: default, global, and temporary configurations, each configurable per-domain. So I would suggest starting from there and trying to find a way to handle these two classes of configuration domains rather than arguing about which specific ones we should include.

**Comment by mlochbaum** [comment]
A few ways to limit the implementation complexity of attributes:
* Only allow attributes for the global configuration. This would mean that per-tab attributes would always override per-domain ones, but I'm starting to be okay with that: if you don't like your per-tab attributes, open a new tab. I'm less okay with session attributes overriding per-domain ones, so that might require a workaround.
* Restrict which options attributes can apply to. For instance, the mode attribute should only apply to ui-based options like color and keybinding while the domain attribute should only apply to content-based options like Javascript settings. This limits some silly-but-slightly useful features like setting the color of the status bar based on the domain.

We already basically have a mode attribute for keybindings, and in fact it might be worth modelling the way per-domain settings work on that existing system. The status bar colors are based on the mode, but not in a principled way, and `hide-statusbar` might be better treated as a per-mode configuration: we set `show-statusbar` to true in prompt and command mode, and when an error or warning is shown (which isn't a mode now but possibly should be), and setting the underlying `show-statusbar` configures when it is shown otherwise.

**Comment by lamarpavel** [comment]
>But we have to have something to store per-domain options, and it seems that there are other attribute-like things that would benefit from the same framework.

See https://github.com/The-Compiler/qutebrowser/issues/27

And I think per-domain settings are much more viable for things like enabling js, cookies etc than per-tab settings. At least that is how it works with all the plugins I have used for different browsers so far.

**Comment by mlochbaum** [comment]
Yes, I would usually set Javascript permissions per-domain. But sometimes there's a particular page on a domain that I want to use Javascript for (an embedded Youtube video, maybe) even though I don't generally want the domain to be Javascript-enabled.

**Comment by Link-Satonaka** [comment]
Couldn't that be handled by userscripts or an autocmd?

**Comment by mlochbaum** [comment]
No? If qutebrowser doesn't support per-tab configuration, I can't set the setting for a single tab.

**Comment by Link-Satonaka** [comment]
It can be done quite easily. An example script would:

disable js when the user accesses a tab or URL with undesired js
enable js when navigating away from said tab or URL

**Comment by mlochbaum** [comment]
So, implement per-tab settings on my own and hope that the site in question doesn't break when I navigate away for a second and Javascript is cut off? Doesn't sound like a solution to me.

**Comment by Link-Satonaka** [comment]
Is JS just cut off, or does that change take effect when you refresh the page?

**Comment by The-Compiler** [comment]
It's cut off.

I can see the need for per-tab settings, and almost everything in [webkitsettings.py](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/browser/webkit/webkitsettings.py#L107-L243) and [webenginesettings.py](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/browser/webengine/webenginesettings.py#L101-L185 should be doable that way. Probably not much of the other settings though.

**Comment by mlochbaum** [comment]
The important step is to make sure that scopes are supported. We can start with pretty much everything disabled for scopes smaller than per-session and then implement per-tab support one option at a time.

**Comment by vyp** [comment]
> Honestly, I'm beginning to wonder again if we really need an .rc file, or if everyone will be able to deal with a simple Python file when it's nicely documented. [...]

@The-Compiler This has probably been mentioned, but just as a reminder, I'm assuming most qutebrowser users are already somewhat familiar with vim, that is to say that a nicely documented python file should not be too 'scary' if that's the concern. Also python is easy and one of the most popular languages too. So big :+1: on this if it reduces complexity and saves time!

**Comment by lahwaacz** [comment]
@mlochbaum All the scopes you propose can be implemented on top of the [The-Compiler's proposal](https://github.com/The-Compiler/qutebrowser/issues/499#issuecomment-204430077), similarly to the example of per-domain settings in that proposal. The problem with your proposal is that it mixes declarative approach (scopes) with imperative approach (attributes) on another level (we already _will_ have declarative `autoconfig.yml` and imperative `config.rc`), while they can all be simply attributes, e.g. of the `:set` command. It might, however, be useful for the internal representation of the config object.

@The-Compiler's quote from [above](https://github.com/The-Compiler/qutebrowser/issues/499#issuecomment-237466613):

> If as an user I have to check ten different places to find out why the heck some option is turned on, that's not a configuration system I want to use. Of course, as a developer, this also makes configuration a lot more difficult to implement and understand.

I think that from the developer's point of view it doesn't matter how many places there are if that's more than 1 :wink: From the user's point of view it's probably similar, except the lower bound might be more than 1. Let's count all the places in the [proposed config](https://github.com/The-Compiler/qutebrowser/issues/499#issuecomment-204430077), assuming that per-domain settings are the only extra scope:

1. Is it a temporary (`:set --temp`) setting?
2. Is it set in `config.py`? (that could be counted as multiple places depending on its complexity)
3. Is it set in `config.rc`? (again could be counted multiple times; the way of setting per-domain options has not been proposed yet)
4. Is it set in `autoconfig.yml` as a global (i.e. not per-domain) option?
5. Is it set in `autoconfig.yml` as a per-domain option? (counted separately because `autoconfig.yml` is a recursive structure)
6. If it's not set anywhere, then it must be the default value.

I think that splitting the configuration into multiple places is the crux of the config revolution and inherently amplified by the separate scope options which have to be supported, like per-domain. Instead of avoiding it, I'd be thinking of providing something like the `which` unix command to print from where does the current value of an option come from.

Also, since `autoconfig.yml` has been proposed as a recursive structure to deal with per-domain settings, I can't think of a sane way to provide a GUI for that. Do you?

**Comment by The-Compiler** [comment]
Wouldn't be something like `:set --domain google.com javascript true` suffice?

**Comment by mlochbaum** [comment]
@lahwaacz I'm not sure why you think per-domain settings are imperative. I view a per-domain setting, such as setting the user agent for Google groups, to be declarative: if the current domain is groups.google.com, then the user agent should be <Chrome UA>. An imperative statement of the same is kind of clunky since it requires us to set the UA on navigating to such a page and unset it on navigating away. Of course, that is what has to happen at the browser level, but I don't think the configuration should have to spell it out.

I believe the stored configuration should always be declarative. Some ways of interacting it, such as `:set` and the rc file, if it exists, use an imperative style, but they are merely issuing a command to change a declarative structure. I didn't write this in my proposal because the terms are somewhat prone to confusion.

**Comment by haasn** [comment]
FWIW I'm a big fan of YAML for user-facing configuration because it allows for very complex configurations (e.g. re-using keys elsewhere).

It's a terrible format for something like wire serialization, but that's not what we're after - is it?

**Comment by rcorre** [comment]
One thing to consider here is writing out only the non-default settings as opposed to all of them. This is more in line with how most config files work.

Right now, if you have `qutebrowser.conf` under version control, you will end up with a large diff of settings you didn't change (and often don't care about) after every update.

I also noticed that setting a single setting with `set` can add a bunch of whitespace to default settings (maybe this is an issue to be fixed separately, but this whole thing can be avoided by not including default settings).

Here's what my diff looks like after using `set` to change the editor:

```diff
diff --git a/qutebrowser/qutebrowser.conf b/qutebrowser/qutebrowser.conf
index caf02b2..fc3e951 100644
--- a/qutebrowser/qutebrowser.conf
+++ b/qutebrowser/qutebrowser.conf
@@ -153,18 +153,18 @@ default-page = ${startpage}
 auto-search = naive
 auto-save-config = true
 auto-save-interval = 15000
-editor = gvim -f "{}"
+editor = gvim -f +"set tw=0" "{}"
 editor-encoding = utf-8
 private-browsing = false
 developer-extras = false
 print-element-backgrounds = true
 xss-auditing = false
 site-specific-quirks = true
-default-encoding =
+default-encoding = 
 new-instance-open-target = window
 log-javascript-console = none
 save-session = true
-session-default-name =
+session-default-name = 
 url-incdec-segments = path,query
 
... and more ...
 
@@ -1454,14 +1454,14 @@ statusbar = 8pt ${_monospace}
 downloads = 8pt ${_monospace}
 hints = bold 12px Monospace
 debug-console = 8pt ${_monospace}
-web-family-standard =
-web-family-fixed =
-web-family-serif =
-web-family-sans-serif =
-web-family-cursive =
-web-family-fantasy =
-web-size-minimum =
-web-size-minimum-logical =
-web-size-default =
-web-size-default-fixed =
+web-family-standard = 
+web-family-fixed = 
+web-family-serif = 
+web-family-sans-serif = 
+web-family-cursive = 
+web-family-fantasy = 
+web-size-minimum = 
+web-size-minimum-logical = 
+web-size-default = 
+web-size-default-fixed = 
 keyhint = 8pt ${_monospace}
```

**Comment by haasn** [comment]
@rcorre The behavior I like is for the default settings to get written commented instead. (And updating qutebrowser would also update the comments to reflect the new values)

**Comment by Link-Satonaka** [comment]
@rcorre @haasn probably half of this thread is me fighting for that "rc config" that vim uses. We came to a compromise- take your time to go through the thread.

**Comment by rcorre** [comment]
@Link-Satonaka, whoops, you're right. I remembered reading this thread months ago and came back to it after running into my config trouble recently. I'll just add that as far as the RC config format, I think `git` is a good example. You can maintain the file by hand and `git config <option> <value>` will try to update the setting in-place. It also has plenty of options that encode 'scopes' into the option names (e.g. `remote.<name>.url).

**Comment by The-Compiler** [comment]
Some abilities the new config should have from an "infrastructure" point of view over the current one:

- Reading settings before verification, so we can have a backend setting
- Being ready as early as possible in the init process - ideally before creating a `QApplication` so we could have a setting for `LIBGL_ALWAYS_SOFTWARE=1` (but I'm not sure the latter is easily possible)
- Marking settings as "needs a restart" with certain backends
- Dynamic reloading of configs
- Reading of config snippets (themes, etc.)

**Comment by throwaway217** [comment]
Since exposing blink/chromium options requires this does this mean this will happen before webengine is default ?

**Comment by The-Compiler** [comment]
Not because of that reason (exposing those options shouldn't block QtWebEngine becoming the default), but it will probably happen before that, or around the same time. Both are planned for v1.0.

**Comment by The-Compiler** [comment]
I just launched qutebrowser's second crowdfunding, with the goal of (finally) making this happen during my summer holidays! Like last year, qutebrowser t-shirts and stickers also are available there.

See the [Kickstarter campaign](https://www.kickstarter.com/projects/the-compiler/qutebrowser-v10-with-per-domain-settings?ref=3kvxwp) for more infos! :tada: :sparkles:

**Comment by Al-Caveman** [comment]
Why don't you use Python as the configuration file? E.g. `qutebrowser.py`, and then define variables just like any python variable, because the config file, itself, is nothing but a python module.  

You can then define loops, arrays, blocks, etc, whatever. Plus, no need to use yet another parser (YAML) in addition to the Python parser.

**Comment by The-Compiler** [comment]
@Al-Caveman please read [the proposal](https://github.com/qutebrowser/qutebrowser/issues/499#issuecomment-204430077) first.

**Comment by mschilli87** [comment]
@The-Compiler: Maybe I missed it but what was the reasoning to drop the `config.rc`?

I really liked the idea.
Not because I consider Python complicated but because I like the idea of using the commands in the config as in command mode (like in `vim`):
Finding an 'interesting' `config.rc` online I could copy/paste it line by line into my active session as command and see the effects.
When I like what I see, I simply add it to my `config.rc`.
This allows me to setup my configuration without the need to even look at the python code (as simple as the 'translation' might be).

I see why it might have lower priority for now, but I really think it would make it easier for less experienced users to get startet with cutomizing their browser.

Also, as far as I can tell `config.load_rc` should be pretty straightforward to implement.

**Comment by The-Compiler** [comment]
More complexity (not code, but documentation, support, etc.) for a little gain, mostly. I'm open to considering it again in the future, but right now I'd rather not have that and see how things go. Less experienced users can still use `qute:settings` or the `:set` command.

**Comment by mschilli87** [comment]
I understand but I think it would be very useful & I would offer to help with the documentation once this is re-considered.

**Comment by noctuid** [comment]
Given how many people wanted the config.rc syntax, I'd argue that a lot of people would find it very useful. It seems to me that it doesn't need much explanation/documentation since the syntax is just a bunch of commands. I also don't think ease of use is really the main issue here. A lot of people have already made good arguments in favor of the config.rc syntax, but I would like to quickly list my reasons for asking for it again:

- It's both intuitive and simple for the user to be able to configure a program in a config file in the same way it is configured while running. The config.rc syntax would just add the ability to persist configuration in its original syntax.
- The method of configuring a program in its commands is common and familiar (pentadactyl, vim, mutt, shell, etc.).
- Many people may never need or want to do anything complicated with python.

For these reasons I think the config.rc is the most natural syntax choice for a configuration file. There is a massive number of people, including myself, who prefer to use handwritten configuration files for everything. There is an overlap between these people and people that might be programmers or would need the config.py, but the two groups are not the same. Plenty of people like to have a handwritten config files not because they necessarily want to do anything complicated in them but because of the other benefits (easy to backup, setup, and edit with a powerful text editor).

I won't argue that the config.py syntax is not simple, but the config.rc syntax is simpler, more expected, and easier to implement. I don't think that either would be a substitute for the other. Given the discussion here, many people would use both. I'd urge you to reconsider this. It could be implemented quickly before the config.py syntax and would be a massive improvement to qutebrowser.  Even after the user could have a config.py, I'd argue that it would still be useful and not really require upkeep.

**Comment by The-Compiler** [comment]
I'm not saying it doesn't have any benefits over a config.py, but I also don't think it's a good idea to have three config files which more or less do the same thing.

**Comment by lahwaacz** [comment]
Given how different qutebrowser's command line is from anything you might be familiar with (see #2017 and the issues linked from there), it's not true that the `config.rc` syntax does not require much explanation. On the other hand, the `config.py` way has a standard syntax so only a documentation of the API would be necessary, which won't be much for the `config.rc` equivalents.

Also, considering that there will be important changes in the commandline parsing, I think it is a good choice to dismiss `config.rc` for the moment and possibly add it later when the syntax is stable.

**Comment by Link-Satonaka** [comment]
I've come around to @The-Compiler 's priorities: do whatever keeps the code and the logic simple, users will adapt to it.

**Comment by Al-Caveman** [comment]
Aftrer reading [this](https://github.com/qutebrowser/qutebrowser/issues/499#issuecomment-204430077), I have some suggestions:


I think it's better to drop this:

> As many errors as possible are catched before python would throw an exception, e.g. arguments are explicitly type-checked. While this is unpythonic, it makes things easier for users not used to python.  

And, instead, adopt a lazy behaviour, where qutebrowser handles errors (e.g. exists) as it reads the configs, without buffering. 

While buffering errors may help a user solve his/her "many" errors a bit faster, such errors happen rarely anyway. Plus, it's unlikely that there are too many nested errors. E.g. usually, we use qutebrowser to browse the Internet. 

Therefore, since errors are not a frequent thing, I think it is not worth optimizing for. So not worth increasing code complexity IMO, even if it's little complexity.

Change:

`
config.set(setting, value, domain=None)
`

To:

`
config.setting(value, group=None)
`
where `setting` is a placeholder for whatever settings that are available for tweaking.

Add:

`
config.url_to_group_mapper(f, group=None)
`

where `f` is a user defined function.

I think it's better to drop this feature:

```
with config.domain('qutebrowser.org'):
        config.set('key1, 'val1')
        config.set('key2, 'val2')
```

IMO, all configurations better consistently follow this single style:

`
config.<setting>(value, group=None)
`

IMO let `group=None` by default, and `config.url_to_group_mapper` by default is configured as follows:
`
config.url_to_group_mapper(f=[lambda url: None], group=None)
`
This way people who don't wish to use per-domain/group settings, they can simply ignore the `group` parameter, and write their configs as follows:
    config.<setting>(value)

I think it's also better to drop all forms of other configurations, and keep only a single one (python).

If anyone wants YAML, RC, etc, then we can add separate modules that update the object `config` by parsing YAML/RC as follows:

`
load_yaml(config, 'file.yml')
`

Or:

`
load_rc(config, 'file.yml')
`

**Comment by The-Compiler** [comment]
> I've said a lot on this subject, so I just want to make it known I've come around to @The-Compiler 's priorities: do whatever keeps the code and the logic simple, users will adapt to it. :)

I wouldn't say that's true, that's more the case with [suckless surf](http://surf.suckless.org/) which in turn isn't really usable 😛 - as mentioned above, I'm mainly worried about the complexity for users there. Or as the [Zen of Python](https://www.python.org/dev/peps/pep-0020/) puts it: "If the implementation is hard to explain, it's a bad idea." :wink:

> > I think it's better to drop this:
> 
> > As many errors as possible are catched before python would throw an exception, e.g. arguments are explicitly type-checked. While this is unpythonic, it makes things easier for users not used to python.  
> 
> And, instead, adopt a lazy behaviour, where qutebrowser handles errors (e.g. exists) as it reads the configs, without buffering. 

This wasn't about buffering errors, but about giving useful error messages.

> Change:
> 
> `
> config.set(setting, value, domain=None)
> `
> 
> To:
> 
> `
> config.setting(value, group=None)
> `
> where `setting` is a placeholder for whatever settings that are available for tweaking.

I actually like the suggestion you made above where the settings are just
attributes. I need to think about that some more. Having them be callables (i.e.
`config.colors.statusbar('#fffff')`) seems odd to me, and not really needed.

> Add:
> 
> `
> config.url_to_group_mapper(f)
> `
> 
> where `f` is a user defined function.
> 
> I think it's better to drop this feature:
> 
> ```
> with config.domain('qutebrowser.org'):
>         config.set('key1, 'val1')
>         config.set('key2, 'val2')
> ```
> 
> IMO, all configurations better consistently follow this single style:
> 
> `
> config.<setting>(value, group=None)
> `
> 
> IMO let `group=None` by default, and `config.url_to_group_mapper` by default is configured as follows:
> `
> config.url_to_group_mapper(f=[lambda url: None])
> `
> This way people who don't wish to use per-domain/group settings, they can simply ignore the `group` parameter, and write their configs as follows:
>     config.<setting>(value)

This sounds much too complicated - I don't see what's wrong with my `with config.domain('qutebrowser.org'):` example above, or what would be better (or simpler) with your proposal.

> I think it's also better to drop all forms of other configurations, and keep only a single one (python).

That means there's no way to configure things via qutebrowser's GUI anymore, and
that's not going to happen. In particular, this also means there wouldn't be a
simple keybinding to toggle e.g. javascript permissions.

> For example, for RC lovers, the `config.py` would have:
> 
> ```python
> from qutebrowser.config.configapi import ConfigAPI
> import load_rc
> config = ConfigAPI()
> load_rc(config, "/path/to/whetever/they/love/their/rc/config/to/be/qute.rc")
> ```

If an RC-like format would be implemented, I also tend towards something like this. However, as said above, I'm not convinced anymore that it's a good idea (and @lahwaacz makes a good point about adding this at a later time, if at all).

**Comment by lahwaacz** [comment]
> I actually like the suggestion you made above where the settings are just
attributes. I need to think about that some more. Having them be callables (i.e.
`config.colors.statusbar('#fffff')`) seems odd to me, and not really needed.

Yep, `config.colors.statusbar = '#ffffff'` is what I'd expect. If other optional arguments for the assignment operator/function are needed, that could be implemented with the context manager syntax: `config.set(setting, value, domain='qutebrowser.org')` would be replaced with
```python
with config.domain('qutebrowser.org'):
    config.setting = value
```

> This sounds much too complicated - I don't see what's wrong with my with `config.domain('qutebrowser.org'):` example above, or what would be better (or simpler) with your proposal.

One notable difference is that @Al-Caveman used `group` instead of `domain`, probably with the intention to match multiple domains at once?

**Comment by Link-Satonaka** [comment]
but suckless is great! And yeah I lazily (incorrectly) worded my response, I do agree with you.

**Comment by noctuid** [comment]
> I'm not saying it doesn't have any benefits over a config.py, but I also don't think it's a good idea to have three config files which more or less do the same thing.

They would each serve a different purpose, and the first is storage only anyway. What exactly is the downside?

> Given how different qutebrowser's command line is from anything you might be familiar with (see #2017 and the issues linked from there), it's not true that the config.rc syntax does not require much explanation or that it is intuitive and simple.

It would not require any additional explanation is my point. The command line needs to be documented whether or not config.rc syntax exists. Issues with the command line are issues with the command line, and whether it should be improved is a separate issue from whether the config.rc syntax should exist. What I was referring to as simple and intuitive is not qutebrowser's command line syntax but allowing a program to be configured in the same way in its configuration file as in its command line. I agree that it is sensible to wait for the command line syntax to stabilize though.

**Comment by Al-Caveman** [comment]
@The-Compiler 
I agree on callables.

Here is currently what I think is best for `config.py` to look like:

```python
# import wants needed
from qutebrowser.config.configapi import ConfigAPI

group[1] = ConfigAPI()
...
group[1].hints.border = "1px solid #E3BE23"
group[1].hints.mode = "number"
... 
group[1].searchengines["gw"] = "https://google.com/q={}"
group[1].aliases["proxysocks"] = "set network proxy socks://127.0.0.1:9050"
...
group[1].ui.keyhint_delay = 0

```



---
@lahwaacz 
> One notable difference is that @Al-Caveman used group instead of domain, probably with the intention to match multiple domains at once?

Yeah. Plus, I don't see why it should be specific "domains", and not more generic forms, such as "URL". Maybe we could do some kinds of regexp, and based on that, treat the URLs (domains being a special case)?

**Comment by Al-Caveman** [comment]
@The-Compiler 
I agree on callables. How about this one, where a setting becomes "per-domain" only if its type is a dictionary:


```python
from qutebrowser.config.configapi import ConfigAPI

config = ConfigAPI()
config.hints.border = "1px solid #E3BE23"
config.hints.mode = {"gmail.com":"number", "yahoo.com":"letters"}
config.searchengines["gw"] = "https://google.com/q={}"
config.aliases["proxysocks"] = "set network proxy socks://127.0.0.1:9050"
config.ui.keyhint_delay = 0
config.command["wo"] = ":open -s {url}"
```

For example, say, `config.hints.border` and `config.command["wo"]` are examples of default settings that applies to all domains. On the other hand, since `type(config.hints.mode)==dict` is `True`, then `config.hints.mode` is a per-domain setting, where its key represents the domain, and its value represents the setting value for that specific domain.

Here is another example, except for using groups (1 and 2) instead of domains:

```python
from qutebrowser.config.configapi import ConfigAPI

config = ConfigAPI()
config.hints.border = "1px solid #E3BE23"
config.hints.mode = {1:"number", 2:"letters"}
config.searchengines["gw"] = "https://google.com/q={}"
config.aliases["proxysocks"] = "set network proxy socks://127.0.0.1:9050"
config.ui.keyhint_delay = 0
config.command["wo"] = ":open -s {url}"
config.domain_to_group_map = f

def f(domain, group):
    if "gmail.com" in domain:
        return 2
    else:
        return 1
```

I think, if there are no technical limitations, it is better to pass URLs instead of domains, so that the domain becomes only a special case of the URL and achieve more flexibility. E.g.

```python
from qutebrowser.config.configapi import ConfigAPI
import re

config = ConfigAPI()
config.hints.border = "1px solid #E3BE23"
config.hints.mode = {1:"number", 2:"letters"}
config.searchengines["gw"] = "https://google.com/q={}"
config.aliases["proxysocks"] = "set network proxy socks://127.0.0.1:9050"
config.ui.keyhint_delay = 0
config.command["wo"] = ":open -s {url}"
config.domain_to_group_map = f

def f(domain, group):
    if re.match(r'^https?://.*?.?lol.com/.*var=popup#test', domain):
        return 2
    else:
        return 1
```


---
@lahwaacz 
> One notable difference is that @Al-Caveman used group instead of domain, probably with the intention to match multiple domains at once?

I don't know what "match multiple domains at once" means. Maybe I accidentally suggested something better without realizing it.

Personally, my reason was based on my intuition that there are more domains than there are "categories of treatment". So it's more convenient for us to define groups of treatments (settings), and then assign the domains (or URLs) to those groups.

My other reason was, once we swallow the idea of groups, another problem is: should we map domains one by one against their groups? I think this is also another inconvenience. This is why I suggested a mapping function, because the function could effectively *generalize* a mappin rule, so that we could potentially put heuristics there, and end up classifying many Internet websites, without having to explecitly list their domains or URLs.

For example, my example above allowed to map any domain based on the domain-agnostic regexp rule: `^https?://.*?.?lol.com/.*var=popup#test`.

**Comment by The-Compiler** [comment]
I'd still rather stay with the context manager syntax (`with config.domain(...):`) and instead allow regexes (or glob patterns, not sure regexes really make sense) there.

Something like:

```
with config.domain(r'/pattern/'):
    config.hints.mode = 'number'
```

or:

```
with config.domain(re.compile(r'pattern')):
    config.hints.mode = 'number'
```

Having a dict as type is already needed for actual setting values, e.g. for `network -> custom-headers`.

**Comment by lahwaacz** [comment]
The problem with regexes and globs is that they increase the complexity of the search. With plain strings like `"qutebrowser.org"` you could store them in a `dict` and have an O(1) search for domains, but with regexes you'd need to iterate over all regexes and for each one check if it matches the given domain/URL, so it's O(_n_). You'd actually need to do this to keep the config general, although in many cases the users' config might have one regex per domain and you'd have no way to optimize for that. I don't know how big the _n_ might be in practice, but it's better to count with enough to make a noticeable difference on the browser's performance. In any case, the _n_ will only get larger over time...

I came upon several ways to solve the matching more efficiently, but the most effective solution is to allow the regexes/globs in `config.domain` as shown above and have some [lru_cache](https://docs.python.org/3/library/functools.html?highlight=lru_cache#functools.lru_cache) for the whole config objects holding the <s>per-domain</s> per-_something_ settings. Also, I don't see the reason to match the full URL, IMO it would be enough to drop the path, query and fragment parts before matching, which would massively increase the efficiency of the cache.

**Comment by noctuid** [comment]
Having both `with config.domain` and `domain=` (and whatever way to match multiple urls/domains) seems perfect to me already.

> not sure regexes really make sense

> Also, I don't see the reason to match the full URL, IMO it would be enough to drop the path

I'd argue in favor of allowing regexes and the ability to match against the url including the path. I don't know about how useful url-specific settings would be, but the vast majority of my pentadactyl keybindings that use groups use globs/regexes to match specific urls. Some keybindings only make sense on specific pages of a website. On github, I match "blob" in the url and change my primary url copying keybinding to first send "y" (to change the url to be for the current commit). For sites that have pages that allow keyboard shortcuts by default (like gmail) or with userscripts, I will pass through certain keys only for those pages. I match certain file extensions (like 'png') to change certain keybindings such as those used to zoom and scroll. This doesn't work for every url that's an image, but it's often useful. The possibilities are endless. Sometimes globs are enough but not always in my experience (and sometimes pentadactyl's limited regex support is less than ideal), so I'd argue for regex support.

**Comment by NoctuaNivalis** [comment]
> The problem with regexes and globs is that they increase the complexity of the search.

It would, but I don't think it'd get too slow. I expect users to have at most 5 or 6 regex domain matchers.

I'd suggest the following syntax:
```python
with config.domain_is('some.domain.com'):
    config.hints.mode = 'number'
with config.domain_matches(re.compile('pattern')):
    config.hints.mode = 'word'
```

As most settings are domain specific, you'd still have the constant time check for settings for the current domain.

**Comment by The-Compiler** [comment]
No need for a separate method - we could simply check whether we get a string or a compiled regex object.

I expect there to be a lot more lookups for per-domain javascript settings, but I guess you'd typically do them via the GUI (with a keybinding) in `autoconfig.yml`, and those can be O(1) (with a lookup per subdomain).

**Comment by Link-Satonaka** [comment]
Actually, I do have one thing to say about configs: whatever you choose, I want to be able to source other config files from within my config.

In my personal dotfiles, I will have a qutebrowserconfig which contains nothing more than my **deviations from default**.

For example,

```
configitem1=true
configitem3=false
source /usr/share/qutebrowser/config
```

Please do not make me import an entire qutebrowser config into my git and then have me track/commit changes as the upstream config file changes.

**Comment by rcorre** [comment]
@Link-Satonaka: based on the [proposal](https://github.com/qutebrowser/qutebrowser/issues/499#issuecomment-204430077), there will no longer be a file pre-populated with all the settings:

> By default, those files are missing/empty, and the built-in defaults get loaded

**Comment by The-Compiler** [comment]
What I haven't really talked about above is how the commands will look to edit some more complex setting types (like lists or dictionaries) from qutebrowser.

Things which used to be a section (like `searchengines`) will now be a dictionary, and the bindings probably will too. There are also various things like `hints -> prev/next-regexes` which will actually be stored as list.

My current idea how to handle this is to add a few new `:config-*` commands, with `:set` being an alias to `:config-set` because `:set` is just a common thing to have.

## Setting

`:config-set` (aka `:set`) will take a JSON syntax to set the elements, and override any previous settings. Examples:
  - `:set custom-headers '{"X-Answer": "42"}'`

## Adding

`:config-add` adds a new element to a dictionary or list.

Examples:
  - `:config-add custom-headers X-Answer 42`
  - `:config-add searchengines wiki https://en.wikipedia.org/wiki/{}`
  - `:config-add next-regexes weiter`

## Removing

`:config-remove` removes an element by key (dicts) or value (lists).

Examples:
  - `:config-remove custom-headers X-Answer`
  - `:config-remove searchengines wiki`
  - `:config-remove next-regexes weiter`

## Replacing

`:config-replace` searches an element as per above, and replaces the value.

Examples:
  - `:config-replace custom-headers X-Answer 23`
  - `:config-replace searchengines wiki https://de.wikipedia.org/wiki/{}`
  - `:config-replace next-regexes weiter [Ww]eiter`

Thoughts? Any other modifications which should be supported?

**Comment by lamarpavel** [comment]
Is the difference between `:config-add` and `:config-replace` just a check for existing entries?

**Comment by The-Compiler** [comment]
@lamarpavel Hmm, good point - I guess it'd be fine for `:config-add` to just replace existing values.

**Comment by lamarpavel** [comment]
Yeah, I think it would be better just to have `:config-add`, maybe with an (optional?) warning for existing entries.

**Comment by The-Compiler** [comment]
Since this will be a breaking change, I'd also like to take the opportunity to reorganize the config to make things clearer - I'd really like some input on that! See #2708.

**Comment by The-Compiler** [comment]
So this issue has gotten way too big - thanks for all your input! I've read through everything again, and then split it into smaller issues:

- YAML config (`autoconfig.yml`): This is now implemented in the `new-config` branch. There'll still be some breaking changes, but it's by all means ready to play with, if you want to test it. It leaves your old config untouched. I'll also add a new post to the [qutebrowser blog](https://www.qutebrowser.org/) with a few more details.
- Python config API: See #2795 (with a completely new API proposal)
- config.rc: See #2796 (my stance is still "probably not going to happen", FWIW)
- Per domain settings: See #27 
- New `:config-*` commands: See #2794 (with some new commands added)
- Per tab/window settings: See #2314

**Comment by The-Compiler** [comment]
Quick update: The `new-config` branch is merged 🎉 A more detailed update will probably follow [in the blog](https://blog.qutebrowser.org/) over the next few days.

**Comment by Cloudef** [comment]
Can't say I'm happy with no tools for migrating the old configuration

**Comment by rcorre** [comment]
There is, if I remember correctly, a page that shows the diff between your config and the default config. It only took me a few minutes to port the diff to the new config, and the manual process helped me learn the new config syntax. I prefer that to having some generated file I don't understand.

**Comment by mschilli87** [comment]
It's `:open qute://configdiff/old`. Also see https://github.com/qutebrowser/qutebrowser/blob/master/doc/help/configuring.asciidoc.

**Comment by The-Compiler** [comment]
An automatic migration also isn't really possible, as things changed in various ways (not only e.g. name changes), and the old config also had default setting values in it (plus defaults changed), so you can't really know automatically what the user intended to change and what not.

**Comment by Cloudef** [comment]
@mschilli87 Yeah, I know about that. It helps, but there's quite a bit of those changes.
@The-Compiler I see, that's a bummer
